src/lib/core/style/add-focus-classes.ts

```diff
@@ -0,0 +1,80 @@
+import {Directive, Injectable, Optional, SkipSelf} from '@angular/core';
+
+
+/** Singleton that allows all instances of CdkAddFocusClasses to share document event listeners. */
+@Injectable()
+export class CdkFocusCauseDetector {
+  /** Whether a keydown event has just occurred. */
+  get keydownOccurred() { return this._keydownOccurred; }
+  private _keydownOccurred = false;
+
+  get mousedownOccurred() { return this._mousedownOccurred; }
+  private _mousedownOccurred = false;
+
+  constructor() {
+    document.addEventListener('keydown', () => {
+      this._keydownOccurred = true;
+      setTimeout(() => this._keydownOccurred = false, 0);
```

Comment:
@DevVersion Why did you guys choose 15ms, that looks kind of race-condition-y.

@jelbourn I could do something similar to the ng1 version I suppose. The scheduled function is about as lightweight as possible, I wouldn't expect it to cost us much in performance. I'll make a little experiment to verify